rust/operator-binary/src/spark_controller.rs

```diff
@@ -0,0 +1,968 @@
+///
+/// This is mostly useful for internal communication between peers, or for clients that perform client-side load balancing.
+fn build_rolegroup_service(
+    sc: &SparkCluster,
+    rolegroup: &RoleGroupRef<SparkCluster>,
+    rolegroup_config: &HashMap<PropertyNameKind, BTreeMap<String, String>>,
+) -> Result<Service, Error> {
+    Ok(Service {
+        metadata: ObjectMetaBuilder::new()
+            .name_and_namespace(sc)
+            .name(&rolegroup.object_name())
+            .ownerreference_from_resource(sc, None, Some(true))
+            .map_err(|e| Error::ObjectMissingMetadataForOwnerRef {
+                source: e,
+                obj_ref: ObjectRef::from_obj(sc),
+            })?
+            .with_recommended_labels(
+                sc,
+                APP_NAME,
+                version(sc)?,
+                &rolegroup.role,
+                &rolegroup.role_group,
+            )
+            .build(),
+        spec: Some(ServiceSpec {
+            cluster_ip: Some("None".to_string()),
+            ports: Some(
+                build_ports(sc, rolegroup, rolegroup_config)
+                    .iter()
+                    .map(|(name, value)| ServicePort {
+                        name: Some(name.clone()),
+                        port: *value,
+                        protocol: Some("TCP".to_string()),
+                        ..ServicePort::default()
+                    })
+                    .collect(),
+            ),
+            selector: Some(role_group_selector_labels(
+                sc,
+                APP_NAME,
+                &rolegroup.role,
+                &rolegroup.role_group,
+            )),
+            publish_not_ready_addresses: Some(true),
+            ..ServiceSpec::default()
+        }),
+        status: None,
+    })
+}
+
+/// Build the [`StatefulSet`]s for the given rolegroup.
+///
+/// # Arguments
+/// * `_sc`                       - The cluster resource object.
+/// * `default_master_role_ports` - Master role service (and container ports). Used to build the master URLs needed by the worker pods.
+/// * `rolegroup`                 - The rolegroup.
+/// * `rolegroup_config`          - The validated configuration for the rolegroup.
+///
+fn build_rolegroup_statefulset(
+    sc: &SparkCluster,
+    default_master_role_ports: &[(String, i32)],
+    rolegroup_ref: &RoleGroupRef<SparkCluster>,
+    rolegroup_config: &HashMap<PropertyNameKind, BTreeMap<String, String>>,
+) -> Result<StatefulSet, Error> {
+    match serde_yaml::from_str(&rolegroup_ref.role).unwrap() {
+        SparkRole::Master => build_master_stateful_set(sc, rolegroup_ref, rolegroup_config),
+        SparkRole::Worker => build_worker_stateful_set(
+            sc,
+            default_master_role_ports,
+            rolegroup_ref,
+            rolegroup_config,
+        ),
+        SparkRole::HistoryServer => build_history_stateful_set(sc, rolegroup_ref, rolegroup_config),
```

Comment:
The `build_*_stateful_set` methods are very similar. Why not implement e.g. command(), args(), log_dir(), container() etc. per role and reuse that method?
![image](https://user-images.githubusercontent.com/19609456/146766842-48e9e22c-0b91-4413-a186-4b3d656c2359.png)

Reply:
I disagree. The methods are 99% the same (and not very short) which adds much more verbosity. We should embrace the benefits of small functions more often, especially with regard to testability.

We always have sets, volume mounts etc. Why not just implement them on a per role basis already?

Changes in the product config should just impact on how we get values, not how we pass them to sets, mounts etc.

Reply:
Ok, I removed most duplicated code.